File: backup/Breadth-first_search(BFS)/c++/03_유기농배추.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/1012
// 2

#include <iostream>
#include <cstring>
#include <queue>

using namespace std;

int matrix[51][51];
bool visited[51][51];
int M, N, K;
int mx[4] = { 1, 0, -1, 0 };
int my[4] = { 0, 1, 0, -1 };

void bfs(int x, int y)
{
	queue<pair<int, int>> q;
	q.push({ x, y });

	while (!q.empty()) {
		pair<int, int> pos = q.front();
		q.pop();

		if (!visited[pos.first][pos.second]) {
			visited[pos.first][pos.second] = true;

			for (int i = 0; i < 4; i++) {
				int nx = pos.first + mx[i];
				int ny = pos.second + my[i];

				if (nx < 0 || nx >= N || ny < 0 || ny >= M)
					continue;

				if (matrix[nx][ny] == 1 && !visited[nx][ny])
					q.push({ nx, ny });
			}
		}
	}
}

int main()
{
	int T;
	cin >> T;

	for (int t = 0; t < T; t++) {
		cin >> M >> N >> K;
		memset(matrix, 0, sizeof(matrix));
		memset(visited, false, sizeof(visited));

		for (int i = 0; i < K; i++) {
			int x, y;
			scanf("%d %d", &x, &y);
			matrix[y][x] = 1;
		}

		int result = 0;
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				if (matrix[i][j] == 1 && !visited[i][j]) {
					bfs(i, j);
					result++;
				}
			}
		}

		cout << result << '\n';
	}
}

File: backup/Breadth-first_search(BFS)/python/03_유기농배추.py
# fast campus 강의
# https://www.acmicpc.net/problem/1012
# 2

from collections import deque

mx = [1, -1, 0, 0]
my = [0, 0, 1, -1]

def bfs(x, y):
    q = deque([x, y])
    while q:
        px = q.popleft()
        py = q.popleft()
        if not(visited[px][py]):
            visited[px][py] = True
            for i in range(4):
                nx = px + mx[i]
                ny = py + my[i]
                if nx < 0 or nx >= n or ny < 0 or ny >= m:
                    continue
                if matrix[nx][ny] and not(visited[nx][ny]):
                    q.append(nx)
                    q.append(ny)

for _ in range(int(input())):
    m, n, k = map(int, input().split())
    matrix = [[0] * m for _ in range(n)]
    visited = [[False] * m for _ in range(n)]

    for _ in range(k):
        x, y = map(int, input().split())
        matrix[y][x] = 1
    
    result = 0
    for i in range(n):
        for j in range(m):
            if matrix[i][j] and not(visited[i][j]):
                bfs(i, j)
                result += 1
    print(result)